File: src/openagents/api/model_provider_api.py
from fastapi import APIRouter, HTTPException, Body

from openagents.repository import model_provider_repository
from openagents.repository.database import ModelProviderEntity

router = APIRouter()


# 查询全部模型提供商，按 id 升序
@router.get("/model-provider/list")
async def list_model_providers() -> list[ModelProviderEntity]:
    return await model_provider_repository.list_model_providers()


# 按 id 查询模型提供商，不存在返回 404
@router.get("/model-provider/{provider_id}")
async def get_model_provider(provider_id: int) -> ModelProviderEntity:
    provider = await model_provider_repository.get_model_provider(provider_id)
    if provider is None:
        raise HTTPException(status_code=404, detail="Model provider not found")
    return provider


# 新增模型提供商，名称已存在返回 409
@router.post("/model-provider")
async def add_model_provider(name: str = Body(..., embed=True), type: str = Body(..., embed=True), base_url: str = Body(..., embed=True), api_key: str = Body(..., embed=True)) -> None:
    if await model_provider_repository.add_model_provider(name, type, base_url, api_key) is None:
        raise HTTPException(status_code=409, detail="Model provider already exists")


# 按 id 更新模型提供商，不存在或名称冲突返回 404
@router.put("/model-provider/{provider_id}")
async def update_model_provider(provider_id: int, name: str = Body(..., embed=True), type: str = Body(..., embed=True), base_url: str = Body(..., embed=True), api_key: str = Body(..., embed=True)) -> None:
    if not await model_provider_repository.update_model_provider(provider_id, name, type, base_url, api_key):
        raise HTTPException(status_code=404, detail="Model provider not found")


# 按 id 删除模型提供商，不存在返回 404，被 Agent 引用返回 409
@router.delete("/model-provider/{provider_id}")
async def delete_model_provider(provider_id: int) -> None:
    if await model_provider_repository.get_model_provider(provider_id) is None:
        raise HTTPException(status_code=404, detail="Model provider not found")
    if not await model_provider_repository.delete_model_provider(provider_id):
        raise HTTPException(status_code=409, detail="Model provider is referenced by agents")

File: src/api/model_provider_api.rs
// 模型提供商 CRUD API
use axum::extract::{Path, State};
use axum::Json;
use serde::Deserialize;

use crate::ai;
use crate::error::AppError;
use crate::repository::entity::ModelProviderEntity;
use crate::repository::{model_provider_repository, DeleteResult};
use crate::state::AppState;

// 查询全部模型提供商, 按 id 升序
pub async fn list_model_providers(
    State(state): State<AppState>,
) -> Result<Json<Vec<ModelProviderEntity>>, AppError> {
    let providers = model_provider_repository::list_model_providers(&state.db).await?;
    Ok(Json(providers))
}

// 按 id 查询模型提供商, 不存在返回 404
pub async fn get_model_provider(
    State(state): State<AppState>,
    Path(provider_id): Path<i64>,
) -> Result<Json<ModelProviderEntity>, AppError> {
    let provider = model_provider_repository::get_model_provider(&state.db, provider_id)
        .await?
        .ok_or_else(|| AppError::NotFound("Model provider not found".to_string()))?;
    Ok(Json(provider))
}

// 模型提供商新增/更新请求体
#[derive(Debug, Deserialize)]
pub struct ModelProviderRequest {
    pub name: String,
    pub protocol_type: String,
    pub base_url: String,
    pub api_key: String,
}

// 新增模型提供商, 返回自增 id
pub async fn add_model_provider(
    State(state): State<AppState>,
    Json(req): Json<ModelProviderRequest>,
) -> Result<Json<i64>, AppError> {
    let id = model_provider_repository::add_model_provider(
        &state.db,
        &req.name,
        &req.protocol_type,
        &req.base_url,
        &req.api_key,
    )
    .await?;
    Ok(Json(id))
}

// 按 id 更新模型提供商, 不存在返回 404
pub async fn update_model_provider(
    State(state): State<AppState>,
    Path(provider_id): Path<i64>,
    Json(req): Json<ModelProviderRequest>,
) -> Result<(), AppError> {
    let updated = model_provider_repository::update_model_provider(
        &state.db,
        provider_id,
        &req.name,
        &req.protocol_type,
        &req.base_url,
        &req.api_key,
    )
    .await?;
    if !updated {
        return Err(AppError::NotFound("Model provider not found".to_string()));
    }
    Ok(())
}

// 按 id 删除模型提供商, 不存在返回 404, 被 Agent 引用返回 409
pub async fn delete_model_provider(
    State(state): State<AppState>,
    Path(provider_id): Path<i64>,
) -> Result<(), AppError> {
    match model_provider_repository::delete_model_provider(&state.db, provider_id).await? {
        DeleteResult::Deleted => Ok(()),
        DeleteResult::NotFound => Err(AppError::NotFound("Model provider not found".to_string())),
        DeleteResult::Referenced => Err(AppError::Conflict(
            "Model provider is referenced by agents".to_string(),
        )),
    }
}

// 查询模型提供商的可用模型列表, 按 provider 协议类型路由实时获取
pub async fn list_provider_models(
    State(state): State<AppState>,
    Path(provider_id): Path<i64>,
) -> Result<Json<Vec<String>>, AppError> {
    let provider = model_provider_repository::get_model_provider(&state.db, provider_id)
        .await?
        .ok_or_else(|| AppError::NotFound("Model provider not found".to_string()))?;
    let models = ai::client::list_models(&provider)
        .await
        .map_err(|e| AppError::Internal(e.into()))?;
    Ok(Json(models))
}
